### ISSUE TITLE
Can't remove channel with nix-channel

### ISSUE THREAD

**Issue by pinage404** [NixOS/nixpkgs#54313]
## Issue description

`nix-shell -p nix-info --run "nix-info -m"` show me a channel that i want to remove

but `nix-channel --list` doesn't show it

and `nix-channel --remove unstable` doesn't remove it

### Steps to reproduce

I don't remember the exact command that i ran, but i guess, it was something like this one

```sh
nix-channel --add https://nixos.org/channels/nixos-unstable unstable
```

Trying to remove it

```
nix-channel --remove unstable
error: opening '/home/pinage404/.nix-channels' for writing: Read-only file system

sudo nix-channel --remove unstable
[sudo] Mot de passe de pinage404 :
```

But the channel is still here

```sh
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.91, NixOS, 18.09.1853.c2950341d03 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(pinage404): `"nixos-18.09.1922.97e0d53d669, nixos-unstable-19.03pre166449.be445a9074f, unstable"`
 - channels(root): `"nixos-18.09.1853.c2950341d03"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

But not everywhere

```sh
nix-channel --list
nixos https://nixos.org/channels/nixos-18.09
nixos-unstable https://nixos.org/channels/nixos-unstable
```

## Technical details

```sh
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.91, NixOS, 18.09.1853.c2950341d03 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(pinage404): `"nixos-18.09.1922.97e0d53d669, nixos-unstable-19.03pre166449.be445a9074f, unstable"`
 - channels(root): `"nixos-18.09.1853.c2950341d03"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by timor** [comment]
I think you executed the original `nix-channel --add` as root.  Did you also check `nix-channel --list` as root?  (I remember being confused by this at first)

**Comment by pinage404** [comment]
Super, you solved my problem !

Thanks !

**Comment by pinage404** [comment]
It came back =(

```sh
# non-root's channels
nix-channel --list
nixos https://nixos.org/channels/nixos-19.03
nixos-unstable https://nixos.org/channels/nixos-unstable

# seems good with my configuration
cat ~/.nix-channels
https://nixos.org/channels/nixos-19.03 nixos
https://nixos.org/channels/nixos-unstable nixos-unstable

# root's channels
sudo nix-channel --list
nixos https://nixos.org/channels/nixos-19.03

sudo nix-channel --remove unstable

nix-channel --remove unstable
error: opening '/home/pinage404/.nix-channels' for writing: Read-only file system

nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.34, NixOS, 19.03.172138.5c52b25283a (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(pinage404): `"nixos-18.09.2505.46d3867a08a, nixos-unstable-19.09pre174426.acbdaa569f4, unstable"`
 - channels(root): `"nixos-19.03.172138.5c52b25283a"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

As you can see the `unstable` channel is come back in my non-root user

How can i remove it ?

I don't know what other informations i should provide

All [my NixOS configuration is here](https://gitlab.com/pinage404/nixos_config) and [my Home Manager configuration is here](https://gitlab.com/pinage404/dotfiles/tree/nixos/config/nixpkgs)

**Comment by timor** [comment]
The name of your channel is `nixos-unstable`, not `unstable` based on what you posted.

**Comment by pinage404** [comment]
this command 
`nix-shell -p nix-info --run "nix-info -m"`

show me `nixos-unstable` and the `unstable` channel

```
 - channels(pinage404): `"nixos-18.09.2505.46d3867a08a, nixos-unstable-19.09pre174426.acbdaa569f4, unstable"`
---------------------------------------------------------------------------------------------------^
```

The real problem is when i search a package using `nox` i have twice the package in the unstable

```
nox vivaldi
Refreshing cache
1 vivaldi-2.3.1440.60-1 (nixos-unstable.vivaldi)
    A Browser for our Friends, powerful and personal
2 vivaldi-ffmpeg-codecs-72.0.3626.122 (nixos-unstable.vivaldi-ffmpeg-codecs)
    Additional support for proprietary codecs for Vivaldi
3 vivaldi-2.3.1440.60-1 (nixos.vivaldi)
    A Browser for our Friends, powerful and personal
4 vivaldi-ffmpeg-codecs-72.0.3626.122 (nixos.vivaldi-ffmpeg-codecs)
    Additional support for proprietary codecs for Vivaldi
5 vivaldi-2.0.1309.29-2 (unstable.vivaldi)
    A Browser for our Friends, powerful and personal
6 vivaldi-ffmpeg-codecs-69.0.3497.73 (unstable.vivaldi-ffmpeg-codecs)
    Additional support for proprietary codecs for Vivaldi
Packages to install:
```

**Comment by timor** [comment]
I don't know what the `nix-info` command output means.  For `nix-channel`, the name before the space in the output of `nix-channel --list` is authoritative as far as I know.

I am unfamiliar with home-manager, but there is a file called `nix-channels` in your config which explicitely lists a `nixos-unstable` channel.  Maybe home-manager will always pull in that channel again because of that?

**Comment by pinage404** [comment]
TL;DR I solved my issue !

---

I had theses channels :
* `nixos` <-- settled by `nix-channels`'s file, i want this
* `nixos-unstable` <-- settled by `nix-channels`'s file, i want this
* `unstable` <-- the one that i don't want

---

### I found that `nix-channels` and `nix-info` don't have the same source of truth

[`nix-channels` uses the `$HOME/.nix-channels`'s file](https://github.com/NixOS/nix/blob/master/src/nix-channel/nix-channel.cc#L167) to know the available channels for the current user

[I use](https://gitlab.com/pinage404/dotfiles/blob/nixos/config/nixpkgs/home.nix#L9) a [`home-manager` feature](https://github.com/rycee/home-manager/blob/4aa07c3547773b798e15666d8599d5650aaf110b/doc/release-notes/rl-1903.xml#L37) to write [the content of this file](https://gitlab.com/pinage404/dotfiles/blob/nixos/config/nixpkgs/nix-channels) (maybe it's not a good idea, i don't know)

In the other hand, [`nix-info` looks in the `/nix/var/nix/profiles/per-user` folder](https://github.com/NixOS/nixpkgs/blob/73debdc298aeb28fe1a86a0b230b34cb11ea5d93/pkgs/tools/nix/info/info.sh#L105)

Where there was the 3 channels : 

```bash
 ll /nix/var/nix/profiles/per-user/pinage404/channels/
total 20
dr-xr-xr-x 2 root root 4096  1 janv.  1970 binary-caches
lrwxrwxrwx 1 root root   60  1 janv.  1970 manifest.nix -> /nix/store/919fi355qd27qdrkhvjxxxja18zx0mvi-env-manifest.nix
lrwxrwxrwx 1 root root   80  1 janv.  1970 nixos -> /nix/store/97i7cgj6bhzk85wrs3pwlav9w7gn2bgh-nixos-19.03.172286.8ea36d73256/nixos
lrwxrwxrwx 1 root root  100  1 janv.  1970 nixos-unstable -> /nix/store/w36wxr1p3f92jzsgiicbnm9cydwhj1y9-nixos-unstable-19.09pre175916.1fc591f9a5b/nixos-unstable
lrwxrwxrwx 1 root root   61  1 janv.  1970 unstable -> /nix/store/k989d02520rrrc99cb39xw9fxiwpysh4-unstable/unstable
```


The file `/nix/var/nix/profiles/per-user/pinage404/channels/manifest.nix` had this content (which i have reformated for GitHub for better readability)

```nix
[
	{
		meta = {};
		name = "nixos-19.03.172286.8ea36d73256";
		out = { outPath = "/nix/store/97i7cgj6bhzk85wrs3pwlav9w7gn2bgh-nixos-19.03.172286.8ea36d73256"; };
		outPath = "/nix/store/97i7cgj6bhzk85wrs3pwlav9w7gn2bgh-nixos-19.03.172286.8ea36d73256";
		outputs = [ "out" ];
		system = "x86_64-linux";
		type = "derivation";
	}
	{
		meta = {};
		name = "nixos-unstable-19.09pre175916.1fc591f9a5b";
		out = { outPath = "/nix/store/w36wxr1p3f92jzsgiicbnm9cydwhj1y9-nixos-unstable-19.09pre175916.1fc591f9a5b"; };
		outPath = "/nix/store/w36wxr1p3f92jzsgiicbnm9cydwhj1y9-nixos-unstable-19.09pre175916.1fc591f9a5b";
		outputs = [ "out" ];
		system = "x86_64-linux";
		type = "derivation";
	}
	{
		meta = {};
		name = "unstable";
		out = { outPath = "/nix/store/k989d02520rrrc99cb39xw9fxiwpysh4-unstable"; };
		outPath = "/nix/store/k989d02520rrrc99cb39xw9fxiwpysh4-unstable";
		outputs = [ "out" ];
		system = "x86_64-linux";
		type = "derivation";
	}
]
```

---

I don't know if i should mention the last commiters to know what should be **the** source of thruth

---

To solves this issue, i removed the `home-manager` configuration that write the `nix-channels` file

```nix
file.".nix-channels".source = ./nix-channels;
```

then i removed the unwanted channel

```sh
nix-channel --remove unstable
```

then i restored my `home-manager` configuration

and now, success, `nix-info --markdown` doesn't give me the `unstable` channel
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.34, NixOS, 19.03.172138.5c52b25283a (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(pinage404): `"nixos-19.03.172286.8ea36d73256, nixos-unstable-19.09pre175916.1fc591f9a5b"`
 - channels(root): `"nixos-19.03.172138.5c52b25283a"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
